File: Arrays/Move_zero_to_end/optimal_sol.cpp
#include <bits/stdc++.h>
using namespace std;

void moveZeroes(int n, vector<int>& a){
    int j = -1;
    for(int i = 0; i<n; i++){
        if(a[i] == 0){
            j = i;
            break;
        }
    }

    // no zero element
    if(j == -1) return;

    for(int i=j+1; i<n; i++){
        if(a[i] != 0){
            swap(a[i],a[j]);
            j++;
        }
    }




}


int main(){
    vector<int> nums = {0,1,0,3,12};
    int n = nums.size();

    moveZeroes(n,nums);

    // Print the result
    for (int num : nums) cout << num << " ";
    cout << endl;
}

File: Arrays/Move_zero_to_end/optimal_sol.py
def moveZeroes(nums):
    n = len(nums)
    i = -1
    for j in range(n):
        if nums[j] ==0:
            i = j
            break

    if i == -1: 
        return nums

    for j in range(i+1,n):
        if nums[j] != 0:
            nums[i], nums[j] = nums[j], nums[i]
            i+= 1

    return nums

if __name__ == "__main__":
    nums = [0,0,0,1,3,4,0,6,7]
    print(moveZeroes(nums))